File: R/Process_NonLongitudinal.r

##We load the utilities
source("Utilities.r")




processNonLogitudinal <- function(cohortName) {
  # The main function to process the NonLongitudinal Data from a cohort
  # Input : cohortName = The name of a valid cohort in cohort builder
  ##Output: barplot showing the availability of the data

  ###First we obtain the patients of the cohort
	patientsCohort <- getCohort(cohortName)
  ##We obtain the data of the non-longitudinal variables
  	dataNonLongitudinal <- getNonLongitudinal(patientsCohort)

  ##We count and summarize the available data
	availableData <- calculateMissing(dataNonLongitudinal, patientsCohort)

  ##We order the available data
  	availableData <- arrange(availableData, desc(percentage))
  ##We clean the variable names
  	availableData <- clean_names(availableData)

  ##We save the table
    write.csv(availableData, "Table_Available_Data_NonLongitudinal.csv",
             row.names = F)

  ##We plot the availableData, the name of the file will be barplotUniqueValues
	plotUniqueValues(availableData, "Percentage_Available_Data_NonLongitudinal")

  }

getNonLongitudinal <- function(patientsCohort) {
  ##Function that obtains and merges all of the tables from non-Longitudinal data
  ##Input: patientsCohort = Vector of patients_id
  ##Output: A dataframe containing all the tables merged
  ######Variables with unique values
	##BirthYear
	birthYear <- getBirthYear(patientsCohort)
	###Employment
	employment <- getEmployment(patientsCohort)
	#Ethnicity
	ethnicity <- getEthnicity(patientsCohort)
	##Gender
	gender <- getGender(patientsCohort)
	##Marital
	marital <- getMarital(patientsCohort)
	##Race
	race <- getRace(patientsCohort)
	##State
	state <- getState(patientsCohort)
	##Education
	education <- getEducationYears(patientsCohort)

	###Merged table for unique values variables
	dfMergedUnique <- merge(birthYear, employment, by = "person_id", all = TRUE)
	dfMergedUnique <- merge(dfMergedUnique, ethnicity, by = "person_id", all = TRUE)
	dfMergedUnique <- merge(dfMergedUnique, gender, by = "person_id", all = TRUE)
	dfMergedUnique <- merge(dfMergedUnique, marital, by = "person_id", all = TRUE)
	dfMergedUnique <- merge(dfMergedUnique, race, by = "person_id", all = TRUE)
	dfMergedUnique <- merge(dfMergedUnique, state, by = "person_id", all = TRUE)
	dfMergedUnique <- merge(dfMergedUnique, education, by = "person_id", all = TRUE)

  	dfMergedUnique <- subset(dfMergedUnique, select=-c(person_id))


  	return(dfMergedUnique)

}

 clean_names <- function(availableData){
 	availableData$variableName <- gsub("years_in_education", "Education years", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("birth_year", "Year of birth", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("ethnicity", "Ethnicity", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("gender", "Gender", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("marital_status", "Marital status", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("state", "State", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("race", "Race", availableData$variableName, fixed=TRUE)
 	availableData$variableName <- gsub("employment_status", "Employment_status", availableData$variableName, fixed=TRUE)
    availableData <- availableData[,c(2,1)]
    colnames(availableData)[1] <- "Variables"
   return(availableData)

 }

File: python/Process_NonLongitudinal.py
# This template demonstrates how to report data availability for non-longitudinal data
# from any cohort in cohort builder.

# Call functions from "Functions.py"
exec(open("Functions.py").read())

def processNonLongitudinal (cohort_name):
  # Get patient IDs stored within a Cohort Builder Project named 'Cohort For Onboarding Task'. 
  print("Get desired cohort from cohort builder")
  cohort_patients = nb.get_cohort(cohort_name) 

  # Get variables with unique values in the cohort and store them in a dataframe
  print("Get data for non longitudinal variables of interest")
  cohort_birth_year = nb.get_birth_year(cohort_patients) #Birth year
  cohort_gender = nb.get_gender(cohort_patients) #Gender
  cohort_city = nb.get_city(cohort_patients) #City
  cohort_education_years = nb.get_education_years(cohort_patients) #Education years
  cohort_employment = nb.get_employment(cohort_patients) #Employment
  cohort_ethnicity = nb.get_ethnicity(cohort_patients) #Ethnicity
  cohort_marital = nb.get_marital(cohort_patients) #Marital status
  cohort_race = nb.get_race(cohort_patients) #Race
  cohort_state = nb.get_state(cohort_patients) #State

  # Create list of dataframes
  all_dataframes = [cohort_birth_year, cohort_gender, cohort_city,
                    cohort_education_years, cohort_employment, cohort_ethnicity,
                    cohort_marital, cohort_race, cohort_state]

  # Merge multiple dataframes into one based on 'person_id'
  data = reduce(lambda  left,right: pd.merge(left,right,on=['person_id'],
                                             how='outer'), all_dataframes)

  # Remove 'person_id' column
  data = data.drop(columns=['person_id'])

  # Replace Unknown / UNKNOWN for Nans
  data = data.replace('Unknown', np.nan)
  data = data.replace('UNKNOWN', np.nan)

  # Create dictionary with neat names for variables
  names={'birth_year':'Year of birth',
         'gender':'Gender',
         'city':'City',
         'years_in_education':'Education years',
         'employment_status':'Employment status',
         'ethnicity':'Ethnicity',
         'marital_status':'Marital status',
         'race':'Race',
         'state':'State'}

  # List of columns names in data
  colnames = list(data.columns)

  # Create table with available data
  dct = {'Percentage_available_data':{}}

  for i in colnames:
    print("Processing %s data"%(i))
    total = 0 # Reset total to zero
    total = len(data['%s'%(i)]) # Estimate total data
    missing = 0 # Reset missing data to zero
    missing = data['%s'%(i)].isnull().sum()
    available = 0 # Reset available to zero
    available = total - missing # Estimate total available data
    per_available = 0 # Reset per_available to zero
    per_available = available / total *100
    # Fill dictionary with calculations
    dct['Percentage_available_data'][names[(i)]] = '%.2f'%(per_available)

  # Transform dictionary into dataframe
  print("Creating dataframe for non-longitudinal data")
  df = pd.DataFrame(dct)
  
  # Convert all columns of DataFrame to numeric
  df = df.apply(pd.to_numeric)
  
  # Change index to a column in df
  df['Variables'] = df.index
  
  # Sort df
  df = df.sort_values(by=['Percentage_available_data'], ascending=True).reset_index(drop=True)
  
  # Reorder columns
  df = df[["Variables", "Percentage_available_data"]]
  
  # Rename Percentage_available_data column for csv file which is created below
  df_nice_names = df.rename(columns={'Percentage_available_data': 'Percentage available (%)'})
  
  # Save df to csv
  print("Percentages were estimated according to the total number of patients, which is equal to %d"%total)
  csv_file_nonlongitudinal = df_nice_names.to_csv('Table_Available_Data_NonLongitudinal.csv')
  
  # Plot Percentage of Data Available 
  print("Creating horizontal barplot for non-longitudinal data")
  hbarplotspercentage_nonlongitudinal(df)
  
  return csv_file_nonlongitudinal
